lib/elixir/lib/calendar.ex

```diff
@@ -99,6 +123,31 @@ defmodule Calendar do
   """
   @callback datetime_to_string(year, month, day, hour, minute, second, microsecond,
                                time_zone, zone_abbr, utc_offset, std_offset) :: String.t
+
+  @doc """
+  Converts the given datetime (with time zone)
+  into the internal Calendar.rata_die format.
+  """
+  @callback datetime_to_rata_die(year, month, day, hour, minute, second, microsecond,
```

Comment:
Who is responsible for handling timezones: datetime or the calendar? I am worried about making it a responsibility of the calendar because it will most likely end-up that all calendars will have to implement the similar timezone conversion logic.

Otherwise we need to specify what is the unity of utc_offset and std_offset. I see advantages in considering them to always be a time fraction of 86400, so we can keep the timezone handling in datetime and then use the calendar module whenever we need to do the offset computation.

For example, if I have 2017-01-01 00:00:00 in ISO and I want to convert it to the Buddhist calendar, it will be converted to the Buddhist calendar in UTC. This means the time fraction from the offsets are not considered in the conversion.

However, if I want to use the Buddhist calendar in Thailand, then I need to convert to Thailand's timezone, which will result in a difference of offsets which will be passed to Thailand's calendar implementation as a time fraction. Then the calendar computes the new year, month, day, hour, minute, second and microsecond.

The advantage of this approach is a simpler API for calendars, as they don't need to worry about timezones.